lib/sparse/src/index/posting_list.rs

```diff
@@ -0,0 +1,189 @@
+use crate::common::types::{DimWeight, RecordId};
+
+#[derive(Debug, Copy, Clone)]
+pub struct PostingElement {
+    pub record_id: RecordId,
+    pub weight: DimWeight,
+    pub max_next_weight: DimWeight,
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct PostingList {
+    /// List of the posting elements ordered by id
+    pub elements: Vec<PostingElement>,
+}
+
+impl PostingList {
+    #[cfg(test)]
+    pub fn from(records: Vec<(RecordId, DimWeight)>) -> PostingList {
+        let mut posting_list = PostingBuilder::new();
+        for (id, weight) in records {
+            posting_list.add(id, weight);
+        }
+        posting_list.build()
+    }
+}
+
+pub struct PostingBuilder {
+    elements: Vec<PostingElement>,
+}
+
+impl PostingBuilder {
+    pub fn new() -> PostingBuilder {
+        PostingBuilder {
+            elements: Vec::new(),
+        }
+    }
+
+    pub fn add(&mut self, record_id: RecordId, weight: DimWeight) {
+        self.elements.push(PostingElement {
+            record_id,
+            weight,
+            max_next_weight: f32::NEG_INFINITY,
+        });
+    }
+
+    pub fn build(mut self) -> PostingList {
+        // Sort by id
+        self.elements.sort_unstable_by_key(|e| e.record_id);
+
+        // Check for duplicates
+        #[cfg(debug_assertions)]
+        {
+            for i in 1..self.elements.len() {
+                if self.elements[i].record_id == self.elements[i - 1].record_id {
+                    panic!(
+                        "Duplicate id {} in posting list",
+                        self.elements[i].record_id
+                    );
+                }
+            }
```

Comment:
_Could_ use `windows`/`find` here:

```suggestion
            if let Some(e) = self
                .elements
                .windows(2)
                .find(|e| e[0].record_id == e[1].record_id)
            {
                panic!("Duplicate id {} in posting list", e[0].record_id);
            }
```